File: sdk-go/files_test.go
package boxkite

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"testing"
)

func TestFileCreate_SendsPathAndContent(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/sandboxes/sess-1/files" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["path"] != "hello.txt" || body["content"] != "hello\n" {
			t.Errorf("unexpected body: %v", body)
		}
		writeJSON(t, w, 200, `{"path": "hello.txt", "size": 6, "created": true}`)
	})
	defer closeServer()

	result, err := client.FileCreate(context.Background(), "sess-1", "hello.txt", "hello\n", nil)
	if err != nil {
		t.Fatalf("FileCreate: %v", err)
	}
	if !result.Created || result.Size != 6 {
		t.Errorf("unexpected result: %+v", result)
	}
}

func TestView_SendsViewRange(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/sandboxes/sess-1/files/view" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		viewRange, _ := body["view_range"].([]any)
		if len(viewRange) != 2 || viewRange[0] != float64(1) || viewRange[1] != float64(10) {
			t.Errorf("unexpected view_range: %v", body["view_range"])
		}
		writeJSON(t, w, 200, `{"content": "hello\n", "lines": 1, "is_directory": false, "entries": null}`)
	})
	defer closeServer()

	result, err := client.View(context.Background(), "sess-1", "hello.txt", &ViewOptions{ViewRange: []int{1, 10}})
	if err != nil {
		t.Fatalf("View: %v", err)
	}
	if result.Content != "hello\n" {
		t.Errorf("unexpected content: %q", result.Content)
	}
}

func TestStrReplace_DefaultsReplaceAllToFalse(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["replace_all"] != false {
			t.Errorf("expected replace_all=false by default, got %v", body["replace_all"])
		}
		if body["old_str"] != "hello" || body["new_str"] != "hi" {
			t.Errorf("unexpected old_str/new_str: %v", body)
		}
		writeJSON(t, w, 200, `{"path": "hello.txt", "replaced": true, "occurrences": 1}`)
	})
	defer closeServer()

	result, err := client.StrReplace(context.Background(), "sess-1", "hello.txt", "hello", "hi", nil)
	if err != nil {
		t.Fatalf("StrReplace: %v", err)
	}
	if result.Occurrences != 1 {
		t.Errorf("unexpected occurrences: %d", result.Occurrences)
	}
}

func TestStrReplace_ReplaceAllTrue(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["replace_all"] != true {
			t.Errorf("expected replace_all=true, got %v", body["replace_all"])
		}
		writeJSON(t, w, 200, `{"path": "hello.txt", "replaced": true, "occurrences": 3}`)
	})
	defer closeServer()

	_, err := client.StrReplace(context.Background(), "sess-1", "hello.txt", "a", "b", &StrReplaceOptions{ReplaceAll: true})
	if err != nil {
		t.Fatalf("StrReplace: %v", err)
	}
}

func TestLs_DefaultsPathToRoot(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["path"] != "/" {
			t.Errorf("expected default path '/', got %v", body["path"])
		}
		writeJSON(t, w, 200, `{"entries": [{"path": "hello.txt", "is_dir": false, "size": 20}]}`)
	})
	defer closeServer()

	result, err := client.Ls(context.Background(), "sess-1", "")
	if err != nil {
		t.Fatalf("Ls: %v", err)
	}
	if len(result.Entries) != 1 || result.Entries[0].Path != "hello.txt" {
		t.Errorf("unexpected entries: %+v", result.Entries)
	}
}

func TestGlob_SendsPatternAndPath(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["pattern"] != "**/*.py" || body["path"] != "/" {
			t.Errorf("unexpected body: %v", body)
		}
		writeJSON(t, w, 200, `{"matches": [{"path": "main.py", "is_dir": false, "size": 120}]}`)
	})
	defer closeServer()

	result, err := client.Glob(context.Background(), "sess-1", "**/*.py", "")
	if err != nil {
		t.Fatalf("Glob: %v", err)
	}
	if len(result.Matches) != 1 {
		t.Errorf("unexpected matches: %+v", result.Matches)
	}
}

func TestGrep_DefaultsMaxMatchesTo500(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["max_matches"] != float64(500) {
			t.Errorf("expected default max_matches=500, got %v", body["max_matches"])
		}
		if _, hasGlob := body["glob"]; hasGlob {
			t.Errorf("expected no glob field when unset, got %v", body["glob"])
		}
		writeJSON(t, w, 200, `{"matches": [{"path": "main.py", "line": 12, "text": "# TODO"}], "error": null, "truncated": false}`)
	})
	defer closeServer()

	result, err := client.Grep(context.Background(), "sess-1", "TODO", nil)
	if err != nil {
		t.Fatalf("Grep: %v", err)
	}
	if len(result.Matches) != 1 || result.Matches[0].Line != 12 {
		t.Errorf("unexpected matches: %+v", result.Matches)
	}
}

func TestGrep_CustomOptions(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["path"] != "/src" || body["glob"] != "*.py" || body["max_matches"] != float64(10) {
			t.Errorf("unexpected body: %v", body)
		}
		writeJSON(t, w, 200, `{"matches": [], "error": null, "truncated": false}`)
	})
	defer closeServer()

	_, err := client.Grep(context.Background(), "sess-1", "TODO", &GrepOptions{Path: "/src", Glob: "*.py", MaxMatches: 10})
	if err != nil {
		t.Fatalf("Grep: %v", err)
	}
}

File: sdk-rust/tests/files_test.rs
mod common;

use boxkite_client::{
    ExecOptions, FileOptions, GlobOptions, GrepOptions, LsOptions, StrReplaceOptions, ViewOptions,
};
use serde_json::json;
use wiremock::matchers::{body_json, method, path};
use wiremock::{Mock, ResponseTemplate};

#[tokio::test]
async fn exec_sends_command_and_parses_result() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess_1/exec"))
        .and(body_json(
            json!({"command": "echo hi", "timeout": 10, "description": "smoke test"}),
        ))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "exit_code": 0, "stdout": "hi\n", "stderr": ""
        })))
        .mount(&server)
        .await;

    let result = client
        .exec(
            "sess_1",
            "echo hi",
            ExecOptions::new().timeout(10).description("smoke test"),
        )
        .await
        .expect("exec should succeed");

    assert_eq!(result.exit_code, 0);
    assert_eq!(result.stdout, "hi\n");
}

#[tokio::test]
async fn exec_omits_optional_fields_when_unset() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess_1/exec"))
        .and(body_json(json!({"command": "echo hi"})))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "exit_code": 0, "stdout": "hi\n", "stderr": ""
        })))
        .mount(&server)
        .await;

    client
        .exec("sess_1", "echo hi", ExecOptions::new())
        .await
        .expect("exec should succeed");
}

#[tokio::test]
async fn file_create_sends_path_and_content() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess_1/files"))
        .and(body_json(json!({"path": "hello.txt", "content": "hi\n"})))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "path": "hello.txt", "size": 3, "created": true
        })))
        .mount(&server)
        .await;

    let result = client
        .file_create("sess_1", "hello.txt", "hi\n", FileOptions::new())
        .await
        .expect("file_create should succeed");

    assert_eq!(result.path, "hello.txt");
    assert_eq!(result.size, 3);
    assert!(result.created);
}

#[tokio::test]
async fn view_sends_view_range() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess_1/files/view"))
        .and(body_json(
            json!({"path": "hello.txt", "view_range": [1, 10]}),
        ))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "content": "hi\n", "lines": 1, "is_directory": false, "entries": null
        })))
        .mount(&server)
        .await;

    let result = client
        .view("sess_1", "hello.txt", ViewOptions::new().view_range(1, 10))
        .await
        .expect("view should succeed");

    assert_eq!(result.content, "hi\n");
    assert!(!result.is_directory);
    assert!(result.entries.is_none());
}

#[tokio::test]
async fn str_replace_sends_replace_all_flag() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess_1/files/str-replace"))
        .and(body_json(json!({
            "path": "hello.txt", "old_str": "hi", "new_str": "hello", "replace_all": true
        })))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "path": "hello.txt", "replaced": true, "occurrences": 2
        })))
        .mount(&server)
        .await;

    let result = client
        .str_replace(
            "sess_1",
            "hello.txt",
            "hi",
            "hello",
            StrReplaceOptions::new().replace_all(true),
        )
        .await
        .expect("str_replace should succeed");

    assert_eq!(result.occurrences, 2);
}

#[tokio::test]
async fn ls_defaults_path_to_root() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess_1/files/ls"))
        .and(body_json(json!({"path": "/"})))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "entries": [{"name": "hello.txt", "is_dir": false}]
        })))
        .mount(&server)
        .await;

    let result = client
        .ls("sess_1", LsOptions::new())
        .await
        .expect("ls should succeed");
    assert_eq!(result.entries.len(), 1);
}

#[tokio::test]
async fn glob_sends_pattern_and_path() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess_1/files/glob"))
        .and(body_json(
            json!({"pattern": "**/*.py", "path": "/workspace"}),
        ))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "matches": [{"path": "/workspace/main.py"}]
        })))
        .mount(&server)
        .await;

    let result = client
        .glob("sess_1", "**/*.py", GlobOptions::new().path("/workspace"))
        .await
        .expect("glob should succeed");

    assert_eq!(result.matches.len(), 1);
}

#[tokio::test]
async fn grep_sends_all_fields() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess_1/files/grep"))
        .and(body_json(json!({
            "pattern": "TODO", "path": "/workspace", "glob": "*.rs", "max_matches": 10
        })))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "matches": [], "error": null, "truncated": false
        })))
        .mount(&server)
        .await;

    let result = client
        .grep(
            "sess_1",
            "TODO",
            GrepOptions::new()
                .path("/workspace")
                .glob("*.rs")
                .max_matches(10),
        )
        .await
        .expect("grep should succeed");

    assert!(result.matches.is_empty());
    assert!(!result.truncated);
}

#[tokio::test]
async fn exec_maps_command_not_allowed_403() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/sandboxes/sess_1/exec"))
        .respond_with(ResponseTemplate::new(403).set_body_json(json!({
            "error": {"code": "command_not_allowed", "message": "This command isn't on the allowlist."}
        })))
        .mount(&server)
        .await;

    let err = client
        .exec("sess_1", "rm -rf /", ExecOptions::new())
        .await
        .unwrap_err();
    assert_eq!(err.code(), Some("command_not_allowed"));
    assert_eq!(err.status(), Some(403));
}
